### PR TITLE
operating system selection added

### LINKED ISSUE #26: os-utils-class
As a user, I want the program to work on an operating system of my choice.

### PR DESCRIPTION
**Pull request by veikkahonkanen**
Resolves #26

### CODE REVIEW

File: classes/os-commands.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  commands: [
+    {name : "copy", command : "cp"}
+  ] 
+},
+{
+  OS : "MAC OS",
+  commands: [
+    {name : "copy", command : "cp"}
+  ] 
+}];
```
Comment (jarinykanen) [inline classes/os-commands.ts]:
```suggestion
const cmds : OsCommand[] = [
  {
    OS : "WINDOWS",
    commands: [
      { name : "copy", command : "copy" }
    ] 
  },
  {
    OS : "LINUX",
    commands: [
      { name : "copy", command : "cp" }
    ] 
  },
  {
    OS : "MAC OS",
    commands: [
      { name : "copy", command : "cp" }
    ] 
  }
];
```

File: classes/os-commands.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+export class OsCommands {
```
Comment (jarinykanen) [inline classes/os-commands.ts:33]: Add docs

File: classes/os-utils.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+/**
+ * Enables cross-platform support
+ * 
+ * getCommand = getter for commands
+ * getOS = getter for current user selected OS 
+ * setOS = setter for user selected OS
+ * 
+ * swapOs = provides functionality for setOS
+ * readUserConfig = provides functionality for getOS and setOS
+ * searchCmd = provides functionality for getCommand
```
Comment (jarinykanen) [inline classes/os-utils.ts]: Move these to function descriptions

Comment (jarinykanen) [reply classes/os-utils.ts]: Also add function params and returns

File: classes/os-utils.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+      } catch (err) {
+        throw err;
+      };
+    }).catch((err) => {
+      throw new Error("Fetching command failed: " + err);
+    });
+    if (!command) {
+      return null;
+    };
+  };
```
Comment (jarinykanen) [inline classes/os-utils.ts]:
```suggestion
  static getCommand = async (cmd: string): Promise<string> => {
    try {
      const userConfig: UserConfigJson = await readUserConfig();
      return searchCmd(userConfig.osPref, cmd);
    } catch (err) {
      return Promise.reject(err);
    };
  };
```

File: classes/os-utils.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        if (osCommands[i].commands[j].name == command) {
+          found = true;
+          return osCommands[i].commands[j].command;
+        };
+      };
+    };
+  };
+  if (!found){
+    return null;
+  };
```
Comment (jarinykanen) [inline classes/os-utils.ts]: These are not used outside this file --> Move inside static class

File: classes/os-utils.ts (hunk 4 above)
Comment (jarinykanen) [reply classes/os-utils.ts]: Use this format in other methods as well.

File: classes/path-utils.ts [hunk 6]
```diff
@@ -32,29 +33,56 @@ export class PathUtils {
...
+ * @param givenPath path that is to be sanitized
+ */
 
-function pathFixer(givenPath : string) : string {
-  
-  if (givenPath[0] === "~") {
-    givenPath = path.join(HOME, givenPath.slice(1))
+function pathFixer(givenPath : string, os? : string) {
+  if (!os) {
+    os = "LINUX";
```
Comment (jarinykanen) [inline classes/path-utils.ts]: Move all OS strings to static enum (include it in the types.ts file)

File: classes/path-utils.ts [hunk 7]
```diff
@@ -32,29 +33,56 @@ export class PathUtils {
...
+/**
+ * Provides cross-platform functionality & tilde expansion when working with paths
+ * @param givenPath path that is to be sanitized
+ */
 
-function pathFixer(givenPath : string) : string {
-  
-  if (givenPath[0] === "~") {
-    givenPath = path.join(HOME, givenPath.slice(1))
+function pathFixer(givenPath : string, os? : string) {
```
Comment (jarinykanen) [inline classes/path-utils.ts]: Change function name to describe what the function actually does. For example translatePath
Add missing params

File: classes/path-utils.ts [hunk 8]
```diff
@@ -32,29 +33,56 @@ export class PathUtils {
...
+  if (!os) {
+    os = "LINUX";
   }
 
-  if (givenPath.match(/^([C-Z]:)/)) {
-    givenPath = path.join(...givenPath.split(/\/|\\/));
+  if (os === "LINUX" || os === "MAC OS") {
+    if (givenPath[0] === "~") {
+      givenPath = path.join(HOME, givenPath.slice(1))
+    };
```
Comment (jarinykanen) [inline classes/path-utils.ts]: Remove all ; after if statements

**Review by jarinykanen — CHANGES_REQUESTED** [review]

File: classes/path-utils.ts [hunk 9]
```diff
@@ -32,29 +34,63 @@ export class PathUtils {
...
-      givenPath = pathFixer(givenPath);
+      let activeOs = null;
+      OsUtils.getOS().then((os) => {
+          activeOs = os;
+      }).catch((err) => {
+
+        if (err) {
+          throw err;
+        }
+      });
```
Comment (anttileppa) [inline classes/path-utils.ts]:
```suggestion
      const activeOs = await OsUtils.getOS();
```

File: classes/path-utils.ts [hunk 10]
```diff
@@ -32,29 +34,63 @@ export class PathUtils {
...
+      if(activeOs){
```
Comment (anttileppa) [inline classes/path-utils.ts]:
```suggestion
      if (activeOs) {
```

File: classes/os-utils.ts [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+    } else {
+      throw new Error("This operating system is not supported.");
+    };
+  };
+};
+
+/**
+ * Functionalities that are used in the exported class above
+ */
+module helperFunctions {
```
Comment (anttileppa) [inline classes/os-utils.ts]: Refactor to private static

File: classes/os-utils.ts [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+  };
+
+  /**
+   * Getter for current user selected OS
+   * 
+   * @returns the selected OS if any
+   */
+  static getOS = async () : Promise<string> => {
+    let os : string = null;
+    await helperFunctions.readUserConfig().then((userConfig : UserConfigJson) => {
```
Comment (anttileppa) [inline classes/os-utils.ts]: Remove callback syntax

File: classes/os-utils.ts [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+import { OsCommands } from "./os-commands";
+import fs from "fs";
+
+const systems : string[]  = ["MAC OS", "WINDOWS", "LINUX"];
+
+/**
+ * Enables cross-platform support
+ * 
+ */
+export class OsUtils {
```
Comment (anttileppa) [inline classes/os-utils.ts]:
```suggestion
export default class OsUtils {
```

File: commands/select-os.ts [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+    });
+
+    if (nameResult.os) {
+      OsUtils.setOS(nameResult.os);
+      
+    } else {
+      throw new Error("ERROR: No operating system has been input.");
+    }
+  } catch (err) {
+    throw(err);
```
Comment (anttileppa) [inline commands/select-os.ts]:
```suggestion
    throw new Error(err);
```

File: classes/path-utils.ts [hunk 15]
```diff
@@ -32,29 +34,63 @@ export class PathUtils {
...
+    if (givenPath[0] === "~") {
+      givenPath = path.join(HOME, givenPath.slice(1))
+    };
+    if (givenPath[0] === "/") {
+      givenPath = path.join(...givenPath.split(/\/|\\/));
+      givenPath = path.sep + givenPath;
+    };
+  }
 
+  if ( os === OperatingSystems.WINDOWS) {
```
Comment (anttileppa) [inline classes/path-utils.ts]:
```suggestion
  if (os === OperatingSystems.WINDOWS) {
```

**Review by anttileppa — CHANGES_REQUESTED** [review]

File: classes/os-commands.ts [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+];
+
+/**
+ * Provides operation system-specific commands for os-utils
+ */
+export class OsCommands {
+  static getCmds = () : OsCommand[] => {
+    return cmds;
+  }
+};
```
Comment (eetupur) [inline classes/os-commands.ts]:
```suggestion
}
```

File: classes/os-utils.ts [hunk 17]
```diff
@@ -0,0 +1,128 @@
+import { OsCommand, CommandObj, UserConfigJson, OperatingSystems } from "../interfaces/types";
+import { OsCommands } from "./os-commands";
+import fs from "fs";
+
+const systems : string[]  = [OperatingSystems.MAC, OperatingSystems.WINDOWS, OperatingSystems.LINUX];
+
+/**
+ * Enables cross-platform support
+ *
```
Comment (eetupur) [inline classes/os-utils.ts]:
```suggestion
```

File: classes/os-utils.ts [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+   * Getter for current user selected OS
+   * 
+   * @returns the selected OS if any
+   */
+  static getOS = async () : Promise<string> => {
+    try{
+        const os : UserConfigJson = await OsUtils.readUserConfig();
+      if (!os) {
+        return null;
+
```
Comment (eetupur) [inline classes/os-utils.ts]:
```suggestion
```

File: classes/os-utils.ts [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+   * @returns either corresponding command or an error
+   */
+  static getCommand = async (cmd: string): Promise<string> => {
+    try {
+      const userConfig: UserConfigJson = await OsUtils.readUserConfig();
+      return OsUtils.searchCmd(userConfig.osPref, cmd);
+    } catch (err) {
+      return Promise.reject(err);
+    };
+  };
```
Comment (eetupur) [inline classes/os-utils.ts]:
```suggestion
    }
  }
```

File: classes/os-utils.ts [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+      if (!os) {
+        return null;
+
+      } else {
+        return os.osPref;
+      }
+    } catch (err) {
+      throw new Error(err);
+    }
+  };
```
Comment (eetupur) [inline classes/os-utils.ts]:
```suggestion
  }
```

File: classes/os-utils.ts [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+
+  /**
+   * Setter for user selected OS
+   * 
+   * @param os is the OS that is being switched to, if supported
+   */
+  static setOS = async (os : string) => {
+    if (systems.includes(os.toUpperCase())) {
+      OsUtils.swapOs(os);
+
```
Comment (eetupur) [inline classes/os-utils.ts]:
```suggestion
```

File: classes/os-utils.ts [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+    } else {
+      throw new Error("This operating system is not supported.");
+    };
+  };
```
Comment (eetupur) [inline classes/os-utils.ts]:
```suggestion
    }
  }
```

File: classes/os-utils.ts [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+    try{
+      fs.writeFile("./user-config.json", data, "utf8", (err) => {
+        if(err){
+          throw new Error("Error when attempting to write file:" + err);
+        }
+      });
+    } catch(err) {
+      throw err;
+    };
+  };
```
Comment (eetupur) [inline classes/os-utils.ts]:
```suggestion
    }
  }
```

File: classes/os-utils.ts [hunk 24]
```diff
@@ -0,0 +1,128 @@
...
+   * 
+   * @param os is the OS that is being switched to, if supported
+   */
+  private static async swapOs (os : string) {
+    const userConfig : UserConfigJson = await OsUtils.readUserConfig();
+    userConfig.osPref = os;
+    let data = JSON.stringify(userConfig, null, 2);
+
+    try{
+      fs.writeFile("./user-config.json", data, "utf8", (err) => {
```
Comment (eetupur) [inline classes/os-utils.ts:67]: Couldn't this be done without callback? For example await.

File: classes/os-utils.ts [hunk 25]
```diff
@@ -0,0 +1,128 @@
...
+   * 
+   * @returns user config as a JSON object that has an interface
+   */
+  private static async readUserConfig() : Promise<UserConfigJson> {
+    let dataJson : UserConfigJson = null;
+    try{
+      let data = fs.readFileSync("./user-config.json", "utf8");
+      dataJson = JSON.parse(data.toString());
+      return dataJson;
+
```
Comment (eetupur) [inline classes/os-utils.ts]:
```suggestion
```

File: classes/os-utils.ts [hunk 26]
```diff
@@ -0,0 +1,128 @@
...
+    let dataJson : UserConfigJson = null;
+    try{
+      let data = fs.readFileSync("./user-config.json", "utf8");
+      dataJson = JSON.parse(data.toString());
+      return dataJson;
+      
+    } catch(err) {
+      throw new Error ("User config not found: " + err);
+    };
+  };
```
Comment (eetupur) [inline classes/os-utils.ts]:
```suggestion
    }
  }
```

File: classes/os-utils.ts [hunk 27]
```diff
@@ -0,0 +1,128 @@
...
+
+  /**
+   * Searches the corresponding command for the corresponding OS
```
Comment (eetupur) [inline classes/os-utils.ts:92]: Not  "...from the correspondin OS"?

File: classes/path-utils.ts [hunk 28]
```diff
@@ -32,29 +34,55 @@ export class PathUtils {
 
   static checkExists = async (givenPath : string) => {
     try {
-      givenPath = pathFixer(givenPath);
+      const activeOs = await OsUtils.getOS();
+      if (activeOs) {
+        givenPath = translatePath(givenPath, activeOs);
+
```
Comment (eetupur) [inline classes/path-utils.ts]:
```suggestion
```

File: classes/path-utils.ts [hunk 29]
```diff
@@ -32,29 +34,55 @@ export class PathUtils {
...
+      } else {
+        givenPath = translatePath(givenPath);
+      }
+
       if (!fs.existsSync(givenPath)) {
         execSync(`mkdir ${givenPath}`);
       }
+
```
Comment (eetupur) [inline classes/path-utils.ts]:
```suggestion
```

File: classes/path-utils.ts [hunk 30]
```diff
@@ -32,29 +34,55 @@ export class PathUtils {
...
     } catch(err) {
       throw new Error(`Error when checking or creating path: ${err}`);
-    }
-  }
-}
+    };
+  };
+};
```
Comment (eetupur) [inline classes/path-utils.ts]:
```suggestion
    }
  }
}
```

File: classes/path-utils.ts [hunk 31]
```diff
@@ -32,29 +34,55 @@ export class PathUtils {
...
-  if (givenPath.match(/^([C-Z]:)/)) {
-    givenPath = path.join(...givenPath.split(/\/|\\/));
+  if (os === OperatingSystems.LINUX || os === OperatingSystems.MAC) {
+    if (givenPath[0] === "~") {
+      givenPath = path.join(HOME, givenPath.slice(1))
+    };
+    if (givenPath[0] === "/") {
+      givenPath = path.join(...givenPath.split(/\/|\\/));
+      givenPath = path.sep + givenPath;
+    };
```
Comment (eetupur) [inline classes/path-utils.ts]:
```suggestion
    }
```

File: classes/path-utils.ts [hunk 32]
```diff
@@ -32,29 +34,55 @@ export class PathUtils {
...
+  if (!os) {
+    os = OperatingSystems.LINUX;
   }
 
-  if (givenPath.match(/^([C-Z]:)/)) {
-    givenPath = path.join(...givenPath.split(/\/|\\/));
+  if (os === OperatingSystems.LINUX || os === OperatingSystems.MAC) {
+    if (givenPath[0] === "~") {
+      givenPath = path.join(HOME, givenPath.slice(1))
+    };
```
Comment (eetupur) [inline classes/path-utils.ts]:
```suggestion
    }
```

File: classes/path-utils.ts [hunk 33]
```diff
@@ -32,29 +34,55 @@ export class PathUtils {
...
+function translatePath(givenPath : string, os? : string) {
+  if (!os) {
+    os = OperatingSystems.LINUX;
   }
 
-  if (givenPath.match(/^([C-Z]:)/)) {
-    givenPath = path.join(...givenPath.split(/\/|\\/));
+  if (os === OperatingSystems.LINUX || os === OperatingSystems.MAC) {
+    if (givenPath[0] === "~") {
+      givenPath = path.join(HOME, givenPath.slice(1))
```
Comment (eetupur) [inline classes/path-utils.ts]:
```suggestion
      givenPath = path.join(HOME, givenPath.slice(1));
```

File: classes/path-utils.ts [hunk 34]
```diff
@@ -32,29 +34,55 @@ export class PathUtils {
...
+    if (givenPath[0] === "/") {
+      givenPath = path.join(...givenPath.split(/\/|\\/));
+      givenPath = path.sep + givenPath;
+    };
+  }
 
+  if (os === OperatingSystems.WINDOWS) {
+    if (givenPath.match(/^([C-Z]:)/)) {
+      givenPath = path.join(...givenPath.split(/\/|\\/));
+    };
```
Comment (eetupur) [inline classes/path-utils.ts]:
```suggestion
    }
```

File: classes/path-utils.ts [hunk 35]
```diff
@@ -32,29 +34,55 @@ export class PathUtils {
...
   } else {
     givenPath = path.join(...givenPath.split(/\/|\\/));
     givenPath = path.sep + givenPath;
   }
 
   return givenPath;
-}
+};
```
Comment (eetupur) [inline classes/path-utils.ts]:
```suggestion
}
```

File: commands/new-repo.ts [hunk 36]
```diff
@@ -96,15 +99,25 @@ async function action(args) {
...
+    finishRepo(repoPath);
   } else {
-    execSync("git init", {cwd : repoPath});
-    execSync(`cp project-config.json ${folderPath}`, {cwd : `.${path.sep}resources`})
-    execSync("git checkout -q -b develop", {cwd : repoPath});
-    execSync(`cp README.md ${repoPath}`, {cwd : `.${path.sep}resources`});
-    execSync(`git add README.md`, {cwd : repoPath});
-    execSync(`git commit -q -m "first commit"`, {cwd : repoPath});
-    execSync(`git push -q origin develop`, {cwd : repoPath});
+    OsUtils.getCommand("copy").then((command : string) => {
```
Comment (eetupur) [inline commands/new-repo.ts]: Can this not be done with await and not .then?

File: commands/pull-proj.ts [hunk 37]
```diff
@@ -88,7 +90,7 @@ async function action() {
     execSync(`mkdir ${folderPath}`);
     execSync(`mkdir ${repoPath}`);
     execSync("git init", {cwd : repoPath});
-    execSync(`cp project-config.json ${folderPath}`, {cwd : `.${path.sep}resources`})
+    execSync(`${copy} project-config.json ${folderPath}`, {cwd : `.${path.sep}resources`})
```
Comment (eetupur) [inline commands/pull-proj.ts]:
```suggestion
    execSync(`${copy} project-config.json ${folderPath}`, {cwd : `.${path.sep}resources`});
```

File: commands/new-repo.ts [hunk 38]
```diff
@@ -96,15 +99,25 @@ async function action(args) {
...
+      copy = command;
+      execSync("git init", {cwd : repoPath});
+      execSync(`${copy} project-config.json ${folderPath}`, {cwd : `.${path.sep}resources`})
```
Comment (eetupur) [inline commands/new-repo.ts]:
```suggestion
      execSync(`${copy} project-config.json ${folderPath}`, {cwd : `.${path.sep}resources`});
```

File: commands/select-os.ts [hunk 39]
```diff
@@ -0,0 +1,36 @@
...
+    const nameResult = await this.prompt({
+      type : "list",
+      name : "os",
+      choices : ["LINUX", "WINDOWS", "MAC OS"],
+      message : "Which operating system are you using? "
+    });
+
+    if (nameResult.os) {
+      OsUtils.setOS(nameResult.os);
+
```
Comment (eetupur) [inline commands/select-os.ts]:
```suggestion
```

**Review by eetupur — COMMENTED** [review]
A lot of ; at ends of functions, ifs and elses, don't think they are necessary.

File: commands/new-repo.ts (hunk 36 above)
Comment (veikkahonkanen) [reply commands/new-repo.ts]: Replaced with await

File: classes/os-utils.ts (hunk 24 above)
Comment (veikkahonkanen) [reply classes/os-utils.ts:67]: No. https://nodejs.org/api/fs.html#fs_fs_writefile_file_data_options_callback

File: classes/os-utils.ts [hunk 40]
```diff
@@ -0,0 +1,124 @@
...
+   * Re-writes user config using previous user config as basis
+   * 
+   * @param os is the OS that is being switched to, if supported
+   */
+  private static async swapOs (os : string) {
+    const userConfig : UserConfigJson = await OsUtils.readUserConfig();
+    userConfig.osPref = os;
+    let data = JSON.stringify(userConfig, null, 2);
+
+    try{
```
Comment (anttileppa) [inline classes/os-utils.ts]:
```suggestion
    try {
```

File: classes/os-utils.ts [hunk 41]
```diff
@@ -0,0 +1,124 @@
...
+      throw new Error(err);
+    }
+  }
+
+  /**
+   * Setter for user selected OS
+   * 
+   * @param os is the OS that is being switched to, if supported
+   */
+  static setOS = async (os : string) => {
```
Comment (anttileppa) [inline classes/os-utils.ts]:
```suggestion
  public static setOS = async (os : string) => {
```

File: classes/os-utils.ts [hunk 42]
```diff
@@ -0,0 +1,124 @@
...
+      return Promise.reject(err);
+    }
+  }
+
+  /**
+   * Getter for current user selected OS
+   * 
+   * @returns the selected OS if any
+   */
+  static getOS = async () : Promise<string> => {
```
Comment (anttileppa) [inline classes/os-utils.ts]:
```suggestion
  public static getOS = async () : Promise<string | null> => {
```

File: classes/os-utils.ts [hunk 43]
```diff
@@ -0,0 +1,124 @@
...
+    try{
+        const os : UserConfigJson = await OsUtils.readUserConfig();
+      if (!os) {
+        return null;
+      } else {
+        return os.osPref;
+      }
```
Comment (anttileppa) [inline classes/os-utils.ts]:
```suggestion
    try {
      const os : UserConfigJson = await OsUtils.readUserConfig();
      return os?.osPref || null; 
```

File: classes/os-utils.ts [hunk 44]
```diff
@@ -0,0 +1,124 @@
...
+   * @param os is the OS that is being switched to, if supported
+   */
+  private static async swapOs (os : string) {
+    const userConfig : UserConfigJson = await OsUtils.readUserConfig();
+    userConfig.osPref = os;
+    let data = JSON.stringify(userConfig, null, 2);
+
+    try{
+      fs.writeFile("./user-config.json", data, "utf8", (err) => {
+        if(err){
```
Comment (anttileppa) [inline classes/os-utils.ts]:
```suggestion
        if (err) {
```

File: classes/os-utils.ts [hunk 45]
```diff
@@ -0,0 +1,124 @@
...
+
+    /**
+   * Re-writes user config using previous user config as basis
+   * 
+   * @param os is the OS that is being switched to, if supported
+   */
+  private static async swapOs (os : string) {
+    const userConfig : UserConfigJson = await OsUtils.readUserConfig();
+    userConfig.osPref = os;
+    let data = JSON.stringify(userConfig, null, 2);
```
Comment (anttileppa) [inline classes/os-utils.ts]:
```suggestion
    const data = JSON.stringify(userConfig, null, 2);
```

File: classes/os-utils.ts [hunk 46]
```diff
@@ -0,0 +1,124 @@
...
+   */
+  private static async readUserConfig() : Promise<UserConfigJson> {
+    let dataJson : UserConfigJson = null;
+    try{
+      let data = fs.readFileSync("./user-config.json", "utf8");
+      dataJson = JSON.parse(data.toString());
+      return dataJson;
+    } catch(err) {
+      throw new Error ("User config not found: " + err);
+    }
```
Comment (anttileppa) [inline classes/os-utils.ts]:
```suggestion
    try {
      const data = fs.readFileSync("./user-config.json", "utf8");
      return JSON.parse(data.toString());
    } catch(err) {
      throw new Error ("User config not found: " + err);
    }
```

File: classes/os-utils.ts [hunk 47]
```diff
@@ -0,0 +1,124 @@
...
+  }
+
+  /**
+   * Searches the corresponding command for the corresponding OS
+   * 
+   * @param os is the OS that is currently being used
+   * 
+   * @param command is the command's name that is being searched
+   */
+  private static searchCmd (os : string, command : string) {
```
Comment (anttileppa) [inline classes/os-utils.ts]:
```suggestion
  private static searchCmd (os: string, command: string) {
```

Comment (anttileppa) [inline classes/os-utils.ts]: Add returns docs

File: classes/os-utils.ts [hunk 48]
```diff
@@ -0,0 +1,124 @@
...
+
+            found = true;
+            return osCommands[i].commands[j].command;
+          }
+        }
+      }
+    }
+    if (!found){
+      return null;
+    }
```
Comment (anttileppa) [inline classes/os-utils.ts]:
```suggestion
    return OsCommands.getCmds()
      .filter(item => item.OS == os)
      .flatMap(item => item.commands)
      .find(item => item.name === command);
```

File: classes/path-utils.ts [hunk 49]
```diff
@@ -32,7 +34,13 @@ export class PathUtils {
 
   static checkExists = async (givenPath : string) => {
     try {
-      givenPath = pathFixer(givenPath);
+      const activeOs = await OsUtils.getOS();
+      if (activeOs) {
+        givenPath = translatePath(givenPath, activeOs);
+      } else {
+        givenPath = translatePath(givenPath);
+      }
```
Comment (anttileppa) [inline classes/path-utils.ts]:
```suggestion
      givenPath = activeOs ? translatePath(givenPath, activeOs) : translatePath(givenPath;
```

File: commands/new-repo.ts [hunk 50]
```diff
@@ -96,15 +99,25 @@ async function action(args) {
...
+    finishRepo(repoPath);
   } else {
-    execSync("git init", {cwd : repoPath});
-    execSync(`cp project-config.json ${folderPath}`, {cwd : `.${path.sep}resources`})
-    execSync("git checkout -q -b develop", {cwd : repoPath});
-    execSync(`cp README.md ${repoPath}`, {cwd : `.${path.sep}resources`});
-    execSync(`git add README.md`, {cwd : repoPath});
-    execSync(`git commit -q -m "first commit"`, {cwd : repoPath});
-    execSync(`git push -q origin develop`, {cwd : repoPath});
+    try{
```
Comment (anttileppa) [inline commands/new-repo.ts]:
```suggestion
    try {
```

**Review by anttileppa — CHANGES_REQUESTED** [review]

File: classes/os-utils.ts [hunk 51]
```diff
@@ -0,0 +1,106 @@
...
+   * Getter for commands
+   * 
+   * @param cmd refers to command name that is being searched
+   * 
+   * @returns either corresponding command or an error
+   */
+  static getCommand = async (cmd: string): Promise<string> => {
+    try {
+      const userConfig: UserConfigJson = await OsUtils.readUserConfig();
+      const command : CommandObj = OsUtils.searchCmd(userConfig.osPref, cmd)
```
Comment (eetupur) [inline classes/os-utils.ts]:
```suggestion
      const command : CommandObj = OsUtils.searchCmd(userConfig.osPref, cmd);
```

File: commands/select-os.ts [hunk 52]
```diff
@@ -0,0 +1,35 @@
...
+import OsUtils from "../classes/os-utils";
+
+const vorpal = new Vorpal();
+
+/**
+ * Prompts the user to input their OS for running other commands later on
+ */
+async function action() {
+  try {
+    const nameResult = await this.prompt({
```
Comment (eetupur) [inline commands/select-os.ts]:
```suggestion
    const osResult = await this.prompt({
```

**Review by eetupur — COMMENTED** [review]
Found a few things that could be improved

File: classes/os-utils.ts (hunk 51 above)
Comment (anttileppa) [inline classes/os-utils.ts]:
```suggestion
      const command : CommandObj = OsUtils.searchCmd(userConfig.osPref, cmd);
```

**Review by anttileppa — CHANGES_REQUESTED** [review]

File: commands/new-repo.ts [hunk 53]
```diff
@@ -30,6 +31,8 @@ async function action(args) {
     args.options.path :
     defaultPath;
 
+  let copy : string = "";
+
```
Comment (anttileppa) [inline commands/new-repo.ts]:
```suggestion
```

File: commands/new-repo.ts [hunk 54]
```diff
@@ -96,15 +99,25 @@ async function action(args) {
...
   } else {
-    execSync("git init", {cwd : repoPath});
-    execSync(`cp project-config.json ${folderPath}`, {cwd : `.${path.sep}resources`})
-    execSync("git checkout -q -b develop", {cwd : repoPath});
-    execSync(`cp README.md ${repoPath}`, {cwd : `.${path.sep}resources`});
-    execSync(`git add README.md`, {cwd : repoPath});
-    execSync(`git commit -q -m "first commit"`, {cwd : repoPath});
-    execSync(`git push -q origin develop`, {cwd : repoPath});
+    try {
+      const copy : string = await OsUtils.getCommand("copy");
```
Comment (anttileppa) [inline commands/new-repo.ts]:
```suggestion
    const copy : string = await OsUtils.getCommand("copy");
    try {
```

File: classes/path-utils.ts [hunk 55]
```diff
@@ -42,15 +46,33 @@ export class PathUtils {
...
-    givenPath = path.join(HOME, givenPath.slice(1))
+/**
+ * Provides cross-platform functionality & tilde expansion when working with paths
+ * 
+ * @param givenPath path that is to be sanitized
+ * 
+ * @param os is the OS that is currently being used
+ */
+
+function translatePath(givenPath : string, os? : string) {
```
Comment (anttileppa) [inline classes/path-utils.ts]: Move to private static

File: classes/path-utils.ts [hunk 56]
```diff
@@ -14,12 +16,12 @@ const defaultProjectPath : string = "~/.meta-proj-cli/projects";
  * repoFolder : gives the path format to the folder that actually holds the project (contains .git)
  */
 export class PathUtils {
```
Comment (anttileppa) [inline classes/path-utils.ts:18]: Add visibitities

File: classes/path-utils.ts [hunk 57]
```diff
@@ -14,12 +16,12 @@ const defaultProjectPath : string = "~/.meta-proj-cli/projects";
...
-  static savePath : string = pathFixer(defaultSavePath);
+  static savePath : string = translatePath(defaultSavePath);
```
Comment (anttileppa) [inline classes/path-utils.ts]: change to getter

File: classes/path-utils.ts [hunk 58]
```diff
@@ -14,12 +16,12 @@ const defaultProjectPath : string = "~/.meta-proj-cli/projects";
...
 
-  static projectPath : string = pathFixer(defaultProjectPath);
+  static projectPath : string = translatePath(defaultProjectPath);
```
Comment (anttileppa) [inline classes/path-utils.ts]: change to getter

**Review by anttileppa — CHANGES_REQUESTED** [review]

**Pull request merged**